base/combinatorics.jl

```diff
@@ -242,7 +242,16 @@ function invperm(p::Union{Tuple{},Tuple{Int},Tuple{Int,Int}})
     isperm(p) || throw(ArgumentError("argument is not a permutation"))
     p  # in dimensions 0-2, every permutation is its own inverse
 end
-invperm(a::Tuple) = (invperm([a...])...,)
+
+function invperm(P::NTuple{N,T}) where {N,T}
+    ntuple(Val(N)) do i
+        for j in P
+            @inbounds P[j]==i && return j
+        end
+        throw(ArgumentError("argument is not a permutation"))
+        return first(P) # DO NOT REMOVE: necessary for type stable code_gen
```

Comment:
Are you sure about this? What do you get if you remove this?

Reply:
The compiler can't know that the above loop must terminate, so this makes sense. But it makes just as much sense to make this an error.

Reply:
I don't get it. There is a `throw` just above this? Can someone spell it out to me with a `@code_warntype` or something.

Reply:
Thank you! i just confirmed, that indeed replacing the return with a throw mantains the code typestable.

Reply:
@KristofferC the compiler can never know if any of the returns in the loop are executed, so it correctly infers that the return value of the closure inside ntuple can be either a `Int` or `nothing`,  this can be rectified by adding a phantom return at the end of the closure with the correct type (but this was very inelegant), or as @StefanKarpinski said that invalid exit point can be made into an error, so the compiler know that the only possible exit point from the closure is withing the loop thus it can again infer the right return type, and keep the code typestable

Reply:
The whole time I have looked at this code, there has been a `throw` above the `return` so it is based on this I have commented.

Reply:
> The whole time I have looked at this code, there has been a `throw` above the `return` so it is based on this I have commented.

i'm very sorry @KristofferC